File: glasses/src/reveal.ts
// Streaming "typewriter" reveal for the session screen's newest transcript
// entry. PURE — no I/O, no Date.now; the App feeds it elapsed time.
//
// The live tail (live.ts) and the 6s poll both deliver a transcript entry's
// FULL current text each update. Showing that whole text the instant it
// arrives makes a still-streaming assistant turn land as a block jump every
// frame. Instead we hold a `shown` character count for the NEWEST entry and
// advance it toward the full length over time, so text that grows a little at
// a time reads as it does in the console — typing in.
//
// The catch the brief calls out: don't slow a real block down with a fake
// stream. When the gap between what's shown and the full text is larger than
// REVEAL_SNAP_CHARS (a big chunk landed at once — a user message, a pasted
// blob, a whole message the transcript only recorded on completion), we snap
// straight to the end instead of typewriting it for seconds. Small deltas
// (real streaming) type; large ones appear immediately.
//
// Only the newest entry is ever partially revealed; everything above it is
// already complete and always rendered in full.

export interface RevealState {
  // uuid of the entry `shown` counts characters into — the newest transcript
  // entry for the focused session. null when there's nothing to reveal.
  entryId: string | null;
  shown: number;
}

// Chars/second the typewriter advances at. Fast enough that a normal
// streaming second (tens of chars) drains smoothly rather than lagging behind
// the source, slow enough to still read as typing.
export const REVEAL_RATE_CPS = 150;
// Backlog past which we stop typing and snap to the end: a chunk this large
// arrived as a block, so mirror that instead of animating it.
export const REVEAL_SNAP_CHARS = 200;

export function emptyReveal(): RevealState {
  return { entryId: null, shown: 0 };
}

// A reveal that shows `len` chars of `entryId` outright — used when entering a
// session so the existing transcript renders in full (it's history, not a
// live stream) and only subsequent growth types in.
export function fullReveal(entryId: string | null, len: number): RevealState {
  return { entryId, shown: entryId === null ? 0 : Math.max(0, len) };
}

export interface RevealOpts {
  rateCps?: number;
  snapChars?: number;
}

// Advance the reveal toward the newest entry (`newestId`, `targetLen` chars)
// given `dtMs` elapsed since the last advance. dtMs = 0 is the "text just
// changed" call: it re-anchors to the newest entry and snaps blocks without
// typing anything yet.
export function advanceReveal(
  prev: RevealState,
  newestId: string | null,
  targetLen: number,
  dtMs: number,
  opts: RevealOpts = {}
): RevealState {
  if (newestId === null || targetLen <= 0) {
    return { entryId: newestId, shown: 0 };
  }
  const rate = opts.rateCps ?? REVEAL_RATE_CPS;
  const snap = opts.snapChars ?? REVEAL_SNAP_CHARS;

  // A new entry starts hidden (typed from 0); the same entry keeps whatever
  // was already shown. Either way, clamp to the current length in case the
  // tail re-truncated shorter than before.
  const base = prev.entryId === newestId ? Math.min(prev.shown, targetLen) : 0;
  const backlog = targetLen - base;
  if (backlog <= 0) return { entryId: newestId, shown: targetLen };
  if (backlog > snap) return { entryId: newestId, shown: targetLen };

  const step = Math.floor((rate * Math.max(0, dtMs)) / 1000);
  return { entryId: newestId, shown: Math.min(targetLen, base + step) };
}

// Nothing left to type for the newest entry — the animation loop can stop.
export function revealComplete(state: RevealState, targetLen: number): boolean {
  return state.entryId === null || targetLen <= 0 || state.shown >= targetLen;
}

File: android/app/src/main/java/com/xerktech/turma/core/Reveal.kt
package com.xerktech.turma.core

import kotlin.math.floor
import kotlin.math.min

/**
 * Typewriter reveal state machine — a pure port of glasses/src/reveal.ts.
 *
 * Tracks a character count into the NEWEST transcript entry only; everything
 * above is always rendered in full. A committed entry lands whole (echoed user
 * message, tool result, a message only recorded on completion) and snaps; a
 * genuine live streaming turn types in character-by-character.
 */
data class RevealState(val entryId: String = "", val shown: Int = 0)

const val REVEAL_RATE_CPS = 150
const val REVEAL_SNAP_CHARS = 200

/**
 * @param prev previous reveal state
 * @param newestId id of the newest entry (or the live turn)
 * @param targetLen displayable length of that newest entry
 * @param dtMs elapsed ms since the last tick (0 = "text just changed" re-anchor)
 * @param live true when [newestId] is the in-progress streaming turn (types),
 *             false for a committed entry (snaps on id change)
 */
fun advanceReveal(
    prev: RevealState,
    newestId: String,
    targetLen: Int,
    dtMs: Long,
    live: Boolean,
): RevealState {
    if (targetLen <= 0 || newestId.isEmpty()) return RevealState(newestId, 0)

    val sameId = newestId == prev.entryId
    // A newly-committed entry (id changed, not a live turn) lands whole.
    if (!sameId && !live) return RevealState(newestId, targetLen)

    val base = if (sameId) min(prev.shown, targetLen) else 0
    val backlog = targetLen - base
    // A big chunk arriving at once mirrors immediately instead of animating.
    if (backlog > REVEAL_SNAP_CHARS) return RevealState(newestId, targetLen)

    val step = floor(REVEAL_RATE_CPS * dtMs / 1000.0).toInt()
    val shown = min(targetLen, base + step)
    return RevealState(newestId, shown)
}

/**
 * The live-turn base char count to carry into [advanceReveal], accounting for
 * the fact that the tmux pane scrape driving the live turn is NOT monotonic:
 * mid-generation it SWAPS between unrelated blocks (prose → a `Bash(…)`/`Read(…)`
 * tool bullet → the next prose), so [newText] often does not continue the slice
 * already revealed. When it doesn't, the reveal must SNAP to the new text rather
 * than keep typing from a stale offset — otherwise the last line "deletes and
 * re-streams over and over" (XERK-19). A port of chat.js `repaint`'s
 * `startsWith(revealed prefix)` check, which Android's length-only
 * [advanceReveal] can't do on its own.
 *
 * @param prevText the live text the previous [prevShown] indexes into
 * @param prevShown chars revealed so far of [prevText]
 * @param newText the newest live capture
 * @return [prevShown] (clamped) when [newText] continues the revealed prefix,
 *         else [newText].length (snap to full)
 */
fun liveRevealBase(prevText: String, prevShown: Int, newText: String): Int {
    val revealed = prevText.take(prevShown.coerceIn(0, prevText.length))
    return if (newText.startsWith(revealed)) prevShown.coerceAtMost(newText.length) else newText.length
}

/** Show an entry outright (entering a session so history renders full). */
fun fullReveal(id: String, len: Int): RevealState = RevealState(id, len)

fun revealComplete(state: RevealState, targetLen: Int): Boolean = state.shown >= targetLen
